### PR TITLE
Aliases predictor threadable

### PR DESCRIPTION
**Pull request by jb-leger**
I had problems with aliases and predictors in command cache.

For example:
```
jben@garance ~ $ ipython                                                                                         
Python 3.7.3rc1 (default, Mar 13 2019, 11:01:15) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import os 
   ...: [os.readlink(f'/proc/self/fd/{i}') for i in range(3)]                                                    
Out[1]: ['/dev/pts/5', '/dev/pts/5', '/dev/pts/5']
```

Ok, we are not threaded (because ipython have a predictor). Let's do a alias:

```
jben@garance ~ $ aliases['ipy']='ipython'                                                                        
jben@garance ~ $ ipy                                                                                             
Python 3.7.3rc1 (default, Mar 13 2019, 11:01:15) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import os 
   ...: [os.readlink(f'/proc/self/fd/{i}') for i in range(3)]                                                    
Out[1]: ['/dev/pts/5', '/dev/pts/6', '/dev/pts/7']
```

Ok, we are threaded. Some strange behavior can occurs.

There is no reason for making a different decision for a alias.

This branch build a predictor by analyzing the alias, and the returned predictor is build based on the predictor of original command (special entry, binary analysis or other).

### CODE REVIEW

File: xonsh/commands_cache.py [hunk 1]
```diff
@@ -232,15 +232,46 @@ def predict_threadable(self, cmd):
...
+        command is an alias, elseif build a predictor based on binary analysis
+        on POSIX, else return predict_true.
+        """
+        # alias stuff
+        if not os.path.isabs(cmd0) and os.sep not in cmd0:
+            alss = getattr(builtins, "aliases", dict())
+            if cmd0 in alss:
+                alias_recursion_limit = 10
+                first_args = []
+                while cmd0 in alss:
```
Comment (gforsyth) [inline xonsh/commands_cache.py]: I think a recursive alias resolver would be a generally useful thing -- what about breaking that out into a separate function and then using it here?

File: xonsh/commands_cache.py [hunk 2]
```diff
@@ -232,15 +232,46 @@ def predict_threadable(self, cmd):
...
     #
 
     def default_predictor(self, name, cmd0):
+        """Default predictor, using predictor from original command if the
+        command is an alias, elseif build a predictor based on binary analysis
+        on POSIX, else return predict_true.
+        """
+        # alias stuff
+        if not os.path.isabs(cmd0) and os.sep not in cmd0:
+            alss = getattr(builtins, "aliases", dict())
```
[the remark this answers is not in the document]
Comment (jb-leger) [reply xonsh/commands_cache.py]: Where `aliases` are get from `builtins` in `CommandCache.all_commands`, the name `alss` was used. I use this name only to respect the same naming convention.

File: xonsh/commands_cache.py [hunk 3]
```diff
@@ -232,15 +232,46 @@ def predict_threadable(self, cmd):
...
+            if cmd0 in alss:
+                alias_recursion_limit = 10
```
[the remark this answers is not in the document]
Comment (jb-leger) [reply xonsh/commands_cache.py]: The only reason is to handle infinite loop in aliases definition. If there is a infinite loop in aliases definition, I think the predictor should not go in infinite loop, so I choose a limit, and I have no reason to choose 10. I will add a comment.

File: xonsh/commands_cache.py [hunk 4]
```diff
@@ -213,11 +213,11 @@ def lazy_is_only_functional_alias(self, name):
             val == (name, True) and self.locate_binary(name, ignore_alias=True) is None
         )
 
-    def predict_threadable(self, cmd):
-        """Predicts whether a command list is able to be run on a background
-        thread, rather than the main thread.
+    def predictor_threadable(self, cmd0):
```
[the remark this answers is not in the document]
Comment (jb-leger) [reply xonsh/commands_cache.py]: I agree. On the other hand, the semantics is clear, and I choose this name to have a good significance of names. I will think to other names.

**Comment by jb-leger** [comment]
I added some comments, and rename function to recognize clearly predictor/predict. Also I separate the aliases stuff in a dedicated function.

**Comment by scopatz** [comment]
Thanks @jb-leger! This is some really neat functionality!

**Pull request merged**